// @wowoengine/sawitdb/src/modules/WAL.js
const fs = require('fs');
const crypto = require('crypto');

/**
 * Write-Ahead Logging (WAL) for SawitDB - OPTIMIZED VERSION
 * Redis-level performance with crash safety
 */
class WAL {
    constructor(dbPath, options = {}) {
        this.dbPath = dbPath;
        this.walPath = `${dbPath}.wal`;
        this.enabled = options.enabled !== false;
        this.syncMode = options.syncMode || 'normal';
        this.checkpointInterval = options.checkpointInterval || 10000; // Increased from 1000

        // OPTIMIZATION: Larger buffers and async fsync
        this.groupCommitMs = options.groupCommitMs || 10; // Group commits within 10ms
        this.bufferSize = options.bufferSize || 1024 * 1024; // 1MB buffer

        this.fd = null;
        this.lsn = 0;
        this.operationCount = 0;
        this.pendingOps = [];
        this.writeBuffer = Buffer.allocUnsafe(this.bufferSize);
        this.bufferOffset = 0;
        this.syncTimer = null;
        this.lastSyncTime = Date.now();

        if (this.enabled) {
            this._init();
            this._startSyncTimer();
        }
    }

    _init() {
        // 1. Synchronous phase for Startup/Recovery check
        if (fs.existsSync(this.walPath)) {
            const fd = fs.openSync(this.walPath, 'r+');
            this.lsn = this._readLastLSN(fd);
            fs.closeSync(fd);
        } else {
            this.lsn = 0;
        }

        // 2. Async Write Stream for runtime
        this.stream = fs.createWriteStream(this.walPath, { flags: 'a' });

        // Capture fd for manual fsync if needed
        this.stream.on('open', (fd) => {
            this.fd = fd;
        });

        // Handle errors
        this.stream.on('error', (err) => {
            console.error('[WAL] Stream error:', err);
        });
    }

    // Helper for _init extraction
    _readLastLSN(fd) {
        const stats = fs.fstatSync(fd);
        if (stats.size === 0) return 0;

        let maxLSN = 0;
        const buffer = Buffer.allocUnsafe(stats.size);
        fs.readSync(fd, buffer, 0, stats.size, 0);

        let offset = 0;
        while (offset < buffer.length) {
            const magic = buffer.readUInt32LE(offset);
            if (magic !== 0x57414C00) break;

            const entrySize = buffer.readUInt32LE(offset + 4);
            const lsn = Number(buffer.readBigUInt64LE(offset + 8));

            if (lsn > maxLSN) maxLSN = lsn;
            offset += entrySize;
        }

        return maxLSN;
    }

    /**
     * OPTIMIZED: Async timer-based flush and fsync
     */
    _startSyncTimer() {
        if (this.syncMode === 'off' || this.syncMode === 'full') return;

        this.syncTimer = setInterval(() => {
            // 1. Write any pending application-side buffer to OS
            if (this.bufferOffset > 0) {
                this._flushBuffer();
            }

            // 2. Force OS to persist to Disk (fsync)
            // We use the raw fd from the stream
            if (this.fd) {
                fs.fsync(this.fd, (err) => {
                    if (err) console.error('[WAL] fsync warning:', err.message);
                });
            }
        }, this.groupCommitMs);
    }

    logOperation(operation, table, pageId, beforeImage, afterImage) {
        if (!this.enabled) return;

        this.lsn++;
        this.operationCount++;

        const opCode = this._getOpCode(operation);
        const entry = this._createLogEntry(this.lsn, opCode, table, pageId, beforeImage, afterImage);

        if (this.syncMode === 'full') {
            // FULL: Force sync immediately (blocking, safest)
            // We bypass stream buffer for FULL to ensure it hits disk now?
            // Actually, mixing stream and raw sync write is bad.
            // For FULL mode, we just write to stream and sync immediately?
            // Stream write is async. 
            // To support 'FULL' (Strict Durability) correctly with Streams, we'd need to write and wait for callback.
            // But logOperation is synchronous in current SawitDB architecture.
            // Trade-off: We write to stream, but we can't block easily without de-optimizing.
            // Fallback for FULL: Use fs.writeSync directly?
            // If we use writeSync, we need a separate FD or ensure stream state is okay.
            // Mixing is risky.
            // Safe approach for FULL: Use _flushBuffer() then explicit sync?
            this._writeToBuffer(entry);
            this._flushBuffer(); // Pushes to stream
            // Force sync on stream fd (might block event loop if we use fsyncSync)
            if (this.fd) {
                try { fs.fsyncSync(this.fd); } catch (e) { }
            }
        } else {
            // NORMAL / OFF
            this._writeToBuffer(entry);
        }

        // Auto checkpoint (less frequent)
        if (this.operationCount >= this.checkpointInterval) {
            this.checkpoint();
        }
    }

    _writeToBuffer(entry) {
        // If buffer is full, flush it to stream
        if (this.bufferOffset + entry.length > this.bufferSize) {
            this._flushBuffer();
        }

        entry.copy(this.writeBuffer, this.bufferOffset);
        this.bufferOffset += entry.length;
    }

    _flushBuffer() {
        if (this.bufferOffset === 0) return;

        // WRITE TO ASYNC STREAM
        const chunk = this.writeBuffer.slice(0, this.bufferOffset);
        this.stream.write(chunk);

        this.bufferOffset = 0;
        this.lastSyncTime = Date.now();
    }

    _getOpCode(operation) {
        const codes = {
            'INSERT': 0x01,
            'UPDATE': 0x02,
            'DELETE': 0x03,
            'CREATE_TABLE': 0x04,
            'DROP_TABLE': 0x05,
            'CHECKPOINT': 0x06
        };
        return codes[operation] || 0x00;
    }

    /**
     * OPTIMIZED: Simplified entry format (no checksums for speed)
     */
    _createLogEntry(lsn, opCode, table, pageId, beforeImage, afterImage) {
        const tableNameBuf = Buffer.alloc(32);
        tableNameBuf.write(table);

        const beforeSize = beforeImage ? beforeImage.length : 0;
        const afterSize = afterImage ? afterImage.length : 0;

        // OPTIMIZATION: Remove checksum (4 bytes saved, faster)
        const entrySize = 4 + 4 + 8 + 1 + 32 + 4 + 4 + 4 + beforeSize + afterSize;
        const entry = Buffer.allocUnsafe(entrySize);

        let offset = 0;
        entry.writeUInt32LE(0x57414C00, offset); offset += 4;
        entry.writeUInt32LE(entrySize, offset); offset += 4;
        entry.writeBigUInt64LE(BigInt(lsn), offset); offset += 8;
        entry.writeUInt8(opCode, offset); offset += 1;
        tableNameBuf.copy(entry, offset); offset += 32;
        entry.writeUInt32LE(pageId, offset); offset += 4;
        entry.writeUInt32LE(beforeSize, offset); offset += 4;
        entry.writeUInt32LE(afterSize, offset); offset += 4;

        if (beforeImage) {
            beforeImage.copy(entry, offset);
            offset += beforeSize;
        }

        if (afterImage) {
            afterImage.copy(entry, offset);
        }

        return entry;
    }

    _writeEntry(entry) {
        if (!this.fd) return;
        fs.writeSync(this.fd, entry);
    }

    _flushPendingOps() {
        // Flush buffer if any
        this._flushBuffer();

        // Clear pending ops array
        this.pendingOps = [];
    }

    checkpoint() {
        if (!this.enabled) return;

        this._flushPendingOps();

        const checkpointEntry = this._createLogEntry(
            ++this.lsn,
            this._getOpCode('CHECKPOINT'),
            '',
            0,
            null,
            null
        );
        this._writeEntry(checkpointEntry);
        if (this.fd) {
            try {
                fs.fsyncSync(this.fd);
            } catch (e) {
                // Ignore invalid fd if stream closed or not ready
            }
        }

        this.operationCount = 0;
    }

    recover() {
        if (!this.enabled || !fs.existsSync(this.walPath)) {
            return [];
        }

        const stats = fs.statSync(this.walPath);
        if (stats.size === 0) return [];

        const fd = fs.openSync(this.walPath, 'r');
        const buffer = Buffer.allocUnsafe(stats.size);
        fs.readSync(fd, buffer, 0, stats.size, 0);
        fs.closeSync(fd);

        const operations = [];
        let offset = 0;

        while (offset < buffer.length) {
            try {
                const magic = buffer.readUInt32LE(offset);
                if (magic !== 0x57414C00) break;

                const entrySize = buffer.readUInt32LE(offset + 4);
                const lsn = Number(buffer.readBigUInt64LE(offset + 8));
                const opCode = buffer.readUInt8(offset + 16);
                const tableName = buffer.toString('utf8', offset + 17, offset + 49).replace(/\0/g, '');
                const pageId = buffer.readUInt32LE(offset + 49);
                const beforeSize = buffer.readUInt32LE(offset + 53);
                const afterSize = buffer.readUInt32LE(offset + 57);

                let dataOffset = offset + 61;
                const beforeImage = beforeSize > 0 ? buffer.slice(dataOffset, dataOffset + beforeSize) : null;
                dataOffset += beforeSize;
                const afterImage = afterSize > 0 ? buffer.slice(dataOffset, dataOffset + afterSize) : null;

                operations.push({
                    lsn,
                    opCode,
                    tableName,
                    pageId,
                    beforeImage,
                    afterImage
                });

                offset += entrySize;
            } catch (e) {
                break;
            }
        }

        return operations;
    }

    truncate() {
        if (!this.enabled) return;

        this._flushBuffer(); // Flush any pending first (async)
        // Note: fs.truncateSync works on path or fd. 
        // With stream, it might be tricky if stream is writing.
        // Best effort:
        try { fs.truncateSync(this.walPath, 0); } catch (e) { }
        this.lsn = 0;
        this.operationCount = 0;
    }

    close() {
        if (!this.enabled) return;

        // Stop sync timer
        if (this.syncTimer) {
            clearInterval(this.syncTimer);
            this.syncTimer = null;
        }

        // Final flush
        this._flushBuffer();

        if (this.stream) {
            this.stream.end();
            this.stream = null;
        }

        if (this.syncMode === 'full' && this.fd) {
            try { fs.closeSync(this.fd); } catch (e) { }
        }

        this.fd = null;
    }
}

module.exports = WAL;
